corpus2alpino/annotators/enrich_lassy.py

```diff
@@ -0,0 +1,114 @@
+from corpus2alpino.log import LogSingleton
+from corpus2alpino.models import Document, Utterance
+
+
+class EnrichLassyAnnotator(Annotator):
+    def __init__(self, enrichment_file: str):
+        """Creates an annotator which will add additional attributes to
+        nodes when they match the properties specified in an enrichment
+        file.
+
+        Arguments:
+            enrichment_file {str} -- Path to a comma-separated enrichment file
+            matching attributes columns should be prefixed with @, assignments
+            without. For example the following CSV would match 
+            <node pos="noun" num="pl" /> and add penn_pos="NNS":
+
+            @pos,@num,penn_pos
+            noun,pl,NNS
+            noun,,NN
+            pron,pl,NNPS
+            pron,,NNP
+            det,,DT
+
+            Empty matchers are skipped, the first match is assigned.
+
+        Raises:
+            Exception: Exception is raised if it could not load or
+            parse the enrichment file
+        """
+        with open(enrichment_file) as enrichment:
+            dialect = csv.Sniffer().sniff(enrichment.read(1024))
+            enrichment.seek(0)
+            reader = csv.reader(enrichment, dialect)
+            matchers = {}
+            assigners = {}
+
+            headers = next(reader)
+            for i, header in enumerate(headers):
+                if header.startswith("@"):
+                    matchers[header[1:]] = i
+                else:
+                    assigners[header] = i
+
+            self.enrichments = cast(List[Enrichment], [])
+
+            for row in reader:
+                for i, value in enumerate(row):
+                    self.enrichments.append(Enrichment(
+                        {
+                            key: row[index] for key, index in matchers.items() if row[index]
+                        },
+                        {
+                            key: row[index] for key, index in assigners.items() if row[index]
+                        }))
+
+    def annotate(self, document: Document):
+        for utterance in document.utterances:
+            if not ANNOTATION_KEY in utterance.annotations:
+                LogSingleton.get().error(
+                    Exception("Lassy annotation missing for: {0}|{1}".format(utterance.id, utterance.text)))
+            else:
+                self.enrich_utterance(utterance)
+
+    def enrich_utterance(self, utterance: Utterance):
+        annotation = utterance.annotations[ANNOTATION_KEY]
+        parse = etree.fromstring(annotation)
+        modified = False
+        for node in parse.iter("node"):
+            for enrichment in self.enrichments:
+                if enrichment.is_match(node):
+                    modified = True
+                    break
+        if modified:
+            utterance.annotations[ANNOTATION_KEY] = etree.tostring(
+                parse, encoding='utf8').decode('utf8')
```

Comment:
Easier on the brain if you just inline the final statement before the `break`, as it saves a state variable, three lines of code and a control flow discontinuity:

```python
          for node in parse.iter("node"):
              for enrichment in self.enrichments:
                  if enrichment.is_match(node):
                      utterance.annotations[ANNOTATION_KEY] = etree.tostring(
                           parse, encoding='utf8').decode('utf8')
                      break
 ```